lib/segment/src/id_tracker/mutable_id_tracker.rs

```diff
@@ -0,0 +1,565 @@
+    internal_to_version: Vec<SeqNumberType>,
+    mappings: PointMappings,
+
+    /// List of point versions pending to be persisted, will be persisted on flush
+    pending_versions: Mutex<Vec<VersionChange>>,
+
+    /// List of point mappings pending to be persisted, will be persisted on flush
+    pending_mappings: Mutex<Vec<MappingChange>>,
+}
+
+impl MutableIdTracker {
+    pub fn open(segment_path: PathBuf) -> OperationResult<Self> {
+        let (mappings_path, versions_path) =
+            (mappings_path(&segment_path), versions_path(&segment_path));
+        let (has_mappings, has_versions) = (mappings_path.is_file(), versions_path.is_file());
+
+        // Warn or error about unlikely or problematic scenarios
+        if !has_mappings && has_versions {
+            debug_assert!(
+                false,
+                "Missing mappings file for ID tracker while versions file exists, storage may be corrupted!",
+            );
+            log::error!(
+                "Missing mappings file for ID tracker while versions file exists, storage may be corrupted!",
+            );
+        }
+        if has_mappings && !has_versions {
+            log::warn!(
+                "Missing versions file for ID tracker, WAL should recover point mappings and versions",
+            );
+        }
+
+        let mut deleted = BitVec::new();
+        let mut internal_to_external: Vec<PointIdType> = Default::default();
+        let mut external_to_internal_num: BTreeMap<u64, PointOffsetType> = Default::default();
+        let mut external_to_internal_uuid: BTreeMap<Uuid, PointOffsetType> = Default::default();
+        let mut internal_to_version: Vec<SeqNumberType> = Default::default();
+
+        if has_mappings {
+            load_mappings(
+                &mappings_path,
+                &mut deleted,
+                &mut internal_to_external,
+                &mut external_to_internal_num,
+                &mut external_to_internal_uuid,
+            )
+            .map_err(|err| {
+                OperationError::service_error(format!("Failed to load ID tracker mappings: {err}"))
+            })?;
+        }
+
+        if has_versions {
+            load_versions(
+                &versions_path,
+                &internal_to_external,
+                &external_to_internal_num,
+                &external_to_internal_uuid,
+                &mut internal_to_version,
+            )
+            .map_err(|err| {
+                OperationError::service_error(format!("Failed to load ID tracker versions: {err}"))
+            })?;
+        }
+
+        #[cfg(debug_assertions)]
+        for (idx, id) in external_to_internal_num.iter() {
+            debug_assert!(
+                internal_to_external[*id as usize] == PointIdType::NumId(*idx),
+                "Internal id {id} is mapped to external id {}, but should be {}",
+                internal_to_external[*id as usize],
+                PointIdType::NumId(*idx)
+            );
+        }
```

Comment:
Should `idx` and `id` variable names be swapped? It's external-ID-to-internal-index mapping, so the key is an `id` and the value is an `idx`?

The code usage seems to be correct (e.g., `internal_to_external[*id as usize]`), just variable naming is confusing (why are we indexing `internal_to_external` with `id`? should we not index with... an index?). 😂

Reply:
In a follow up PR I've improved this bit and move it into a function belonging to `PointMappings` because it is used in multiple places:

```rust
impl PointMappings {
    #[cfg(debug_assertions)]
    pub fn assert_mappings(&self) {
        for (external_id, internal_id) in self.external_to_internal_num.iter() {
            debug_assert!(
                self.internal_to_external[*internal_id as usize]
                    == PointIdType::NumId(*external_id),
                "Internal id {internal_id} is mapped to external id {}, but should be {}",
                self.internal_to_external[*internal_id as usize],
                PointIdType::NumId(*external_id),
            );
        }
    }
}
```